tests/core/pyspec/eth2spec/test/capella/block_processing/test_process_withdrawals.py

```diff
@@ -1,39 +1,76 @@
+import random
+
+from eth2spec.test.context import (
+    spec_state_test,
+    expect_assertion_error,
+    with_capella_and_later,
+    with_presets,
+)
+from eth2spec.test.helpers.constants import MINIMAL
 from eth2spec.test.helpers.execution_payload import (
     build_empty_execution_payload,
 )
+from eth2spec.test.helpers.random import (
+    randomize_state,
+)
+from eth2spec.test.helpers.state import (
+    next_epoch,
+    next_slot,
+)
+from eth2spec.test.helpers.withdrawals import (
+    set_eth1_withdrawal_credential_with_balance,
+    set_validator_fully_withdrawable,
+    set_validator_partially_withdrawable,
+)
 
-from eth2spec.test.context import spec_state_test, expect_assertion_error, with_capella_and_later
 
-from eth2spec.test.helpers.state import next_slot
+def prepare_expected_withdrawals(spec, state,
+                                 num_full_withdrawals=0, num_partial_withdrawals=0, rng=random.Random(5566)):
+    assert num_full_withdrawals + num_partial_withdrawals <= len(state.validators)
+    all_validator_indices = list(range(len(state.validators)))
+    sampled_indices = rng.sample(all_validator_indices, num_full_withdrawals + num_partial_withdrawals)
+    fully_withdrawable_indices = rng.sample(sampled_indices, num_full_withdrawals)
+    partial_withdrawals_indices = list(set(sampled_indices).difference(set(fully_withdrawable_indices)))
 
+    for index in fully_withdrawable_indices:
+        set_validator_fully_withdrawable(spec, state, index)
+    for index in partial_withdrawals_indices:
+        set_validator_partially_withdrawable(spec, state, index)
 
-def prepare_withdrawal_queue(spec, state, num_withdrawals):
-    pre_queue_len = len(state.withdrawal_queue)
-    validator_len = len(state.validators)
-    for i in range(num_withdrawals):
-        withdrawal = spec.Withdrawal(
-            index=i + 5,
-            validator_index=(i + 1000) % validator_len,
-            address=b'\x42' * 20,
-            amount=200000 + i,
-        )
-        state.withdrawal_queue.append(withdrawal)
+    return fully_withdrawable_indices, partial_withdrawals_indices
 
-    assert len(state.withdrawal_queue) == num_withdrawals + pre_queue_len
+
+def verify_post_state(state, spec, expected_withdrawals,
+                      fully_withdrawable_indices, partial_withdrawals_indices):
+    expected_withdrawals_validator_indices = [withdrawal.validator_index for withdrawal in expected_withdrawals]
+    assert state.next_withdrawal_index == expected_withdrawals[-1].index + 1
+    assert state.latest_withdrawal_validator_index == expected_withdrawals_validator_indices[-1]
```

Comment:
the tests are passing so we aren't hitting this condition now but this code will blow up if `expected_withdrawals` is empty -- which should be a setting we are testing...

me (or someone :) ) should make a pass on testing anyway once we have agreed on high-level design so im not too worried about this now but I'll call it out

Reply:
true, in all tests that we expect zero withdrawals we don't call this function, but I agree that it would be better to have an API that allows to test this situation as well. I'm adding a return early just in case, and leaving a comment to this review

Reply:
yeah I think we can just direct some more resources in the future to testing this more thoroughly